src/Symfony/Component/PropertyInfo/Util/LegacyTypeConverter.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\PropertyInfo\Util;
+
+use Symfony\Component\PropertyInfo\Type as LegacyType;
+use Symfony\Component\TypeInfo\Type;
+
+/**
+ * @internal
+ */
+class LegacyTypeConverter
+{
+    /**
+     * @param LegacyType[]|null $legacyTypes
+     */
+    public static function toTypeInfoType(?array $legacyTypes): ?Type
+    {
+        if (null === $legacyTypes) {
+            return null;
+        }
+
+        $types = [];
+
+        foreach ($legacyTypes as $legacyType) {
+            $typeInfoType = match ($legacyType->getBuiltinType()) {
+                LegacyType::BUILTIN_TYPE_ARRAY => Type::array(),
+                LegacyType::BUILTIN_TYPE_BOOL => Type::bool(),
+                LegacyType::BUILTIN_TYPE_CALLABLE => Type::callable(),
+                LegacyType::BUILTIN_TYPE_FALSE => Type::false(),
+                LegacyType::BUILTIN_TYPE_FLOAT => Type::float(),
+                LegacyType::BUILTIN_TYPE_INT => Type::int(),
+                LegacyType::BUILTIN_TYPE_ITERABLE => Type::iterable(),
+                LegacyType::BUILTIN_TYPE_NULL => Type::null(),
+                LegacyType::BUILTIN_TYPE_OBJECT => Type::object($legacyType->getClassName()),
+                LegacyType::BUILTIN_TYPE_RESOURCE => Type::resource(),
+                LegacyType::BUILTIN_TYPE_STRING => Type::string(),
+                LegacyType::BUILTIN_TYPE_TRUE => Type::true(),
+            };
+
+            if ($legacyType->isNullable()) {
+                $typeInfoType = Type::union($typeInfoType, Type::null());
```

Comment:
instead of potentially nesting union types (if we have a nullable legacy type in a list with multiple values), and potentially even several ones (if we have multiple nullable types), shouldn't we rather add `Type::null()` in `$types` separately (and maybe avoiding duplicates for it) ?

Reply:
Indeed, that's a valid point. I have changed the code around the handling of `null`.